gcs/constraints/smart_table.cc

```diff
@@ -0,0 +1,642 @@
+    WeightedPseudoBooleanTerms var_1_minus_var_2 = {{1_i, var_1}, {-1_i, var_2}};
+    WeightedPseudoBooleanTerms var_2_minus_var_1 = {{1_i, var_2}, {-1_i, var_1}};
+    switch (c) {
+    case EQUAL:
+        // f => var1 == var2
+        flag = propagators.create_proof_flag("bin_eq");
+        propagators.define_pseudoboolean_eq(state, move(var_1_minus_var_2), 0_i, flag);
+
+        // !f => var1 != var2
+        flag_lt = propagators.create_proof_flag("lt");
+        flag_gt = propagators.create_proof_flag("gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, ! flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, flag_lt);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, ! flag_lt);
+        propagators.define_pseudoboolean_ge(state, {{1_i, flag_lt}, {1_i, flag_gt}}, 1_i, ! flag);
+        return flag;
+    case GREATER_THAN:
+        flag = propagators.create_proof_flag("bin_gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, ! flag);
+        return flag;
+    case LESS_THAN:
+        flag = propagators.create_proof_flag("bin_lt");
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, ! flag);
+        return flag;
+    case LESS_THAN_EQUAL:
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, ! flag);
+        return flag;
+    case NOT_EQUAL:
+        // !f => var1 == var2
+        flag = propagators.create_proof_flag("bin_eq");
+        propagators.define_pseudoboolean_eq(state, move(var_1_minus_var_2), 0_i, ! flag);
+
+        // f => var1 != var2
+        flag_lt = propagators.create_proof_flag("lt");
+        flag_gt = propagators.create_proof_flag("gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 1_i, flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 0_i, ! flag_gt);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, flag_lt);
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, ! flag_lt);
+        propagators.define_pseudoboolean_ge(state, {{1_i, flag_lt}, {1_i, flag_gt}}, 1_i, flag);
+        return flag;
+    case GREATER_THAN_EQUAL:
+        flag = propagators.create_proof_flag("bin_gt");
+        propagators.define_pseudoboolean_ge(state, move(var_1_minus_var_2), 0_i, flag);
+        propagators.define_pseudoboolean_ge(state, move(var_2_minus_var_1), 1_i, ! flag);
+        return flag;
+    default:
```

Comment:
I try to avoid using default in a switch, because it turns off compiler warnings for missing case entries. I would move this outside the switch instead.

Reply:
Also I think there's a specific NonExhaustiveSwitch exception I made for this kind of thing?

Reply:
Fixed in c62e8914eb9e01df2a147f01713ff505d7e905e0